sled-diagnostics/src/contract.rs

```diff
@@ -0,0 +1,172 @@
+pub enum ContractError {
+    #[error(transparent)]
+    FileIo(#[from] std::io::Error),
+    #[error(
+        "Failed to call ct_pr_status_get_svc_fmri for contract {ctid}: {error}"
+    )]
+    Fmri { ctid: i32, error: std::io::Error },
+    #[error(
+        "Failed to call ct_pr_status_get_members for contract {ctid}: {error}"
+    )]
+    Members { ctid: i32, error: std::io::Error },
+    #[error("ct_status_read returned successfully but handed back a null ptr for {0}")]
+    Null(std::path::PathBuf),
+    #[error("Failed to call ct_status_read on {path}: {error}")]
+    StatusRead { path: std::path::PathBuf, error: std::io::Error },
+}
+
+pub struct ContractStatus {
+    handle: ct_stathdl_t,
+}
+
+impl Drop for ContractStatus {
+    fn drop(&mut self) {
+        unsafe { ct_status_free(self.handle) };
+    }
+}
+
+macro_rules! libcall_io {
+        ($fn: ident ( $($arg: expr), * $(,)*) ) => {{
+            let res = unsafe { $fn($($arg, )*) };
+            if res == 0 {
+                Ok(res)
+            } else {
+                Err(std::io::Error::last_os_error())
+            }
+        }};
+    }
+
+impl ContractStatus {
+    fn new(contract_status: &Path) -> Result<Self, ContractError> {
+        let file = fs::File::open(contract_status)?;
+        let mut handle: ct_stathdl_t = std::ptr::null_mut();
+        libcall_io!(ct_status_read(file.as_raw_fd(), CTD_ALL, &mut handle,))
+            .map_err(|error| ContractError::StatusRead {
+                path: contract_status.to_path_buf(),
+                error,
+            })?;
+
+        // We don't ever expect the system to hand back a null ptr when
+        // returning success but let's be extra cautious anyways.
+        if handle.is_null() {
+            return Err(ContractError::Null(contract_status.to_path_buf()));
+        }
+
+        Ok(Self { handle })
+    }
+
+    fn get_members(&self) -> Result<&[i32], ContractError> {
+        let mut numpids = 0;
+        let mut pids: *mut pid_t = std::ptr::null_mut();
+
+        let pids = {
+            libcall_io!(ct_pr_status_get_members(
+                self.handle,
+                &mut pids,
+                &mut numpids,
+            ))
+            .map_err(|error| {
+                let ctid = unsafe { ct_status_get_id(self.handle) };
+                ContractError::Members { ctid, error }
+            })?;
+
+            unsafe {
+                if pids.is_null() {
+                    &[]
+                } else {
+                    std::slice::from_raw_parts(pids, numpids as usize)
+                }
+            }
+        };
+
+        Ok(pids)
+    }
+
+    fn get_fmri(&self) -> Result<Option<CString>, ContractError> {
+        // The lifetime of this string is tied to the lifetime of the status
+        // handle itself and will be cleaned up when the handle is freed.
+        let mut ptr: *mut c_char = std::ptr::null_mut();
+        libcall_io!(ct_pr_status_get_svc_fmri(self.handle, &mut ptr)).map_err(
```

Comment:
From https://illumos.org/man/3CONTRACT/ct_pr_status_get_svc_creator#:~:text=The%20ct_pr_status_get_svc_fmri(),not%20be%20modified

>       The ct_pr_status_get_svc_fmri(), ct_pr_status_get_svc_creator(), and
>       ct_pr_status_get_svc_aux() functions read, respectively, the service
>       FMRI, the contract's creator execname and  the creator's auxiliary field.
>       The buffer pointed to by fmri, aux or creator, is freed by a call to
>       ct_status_free() and should not be modified.


I'm under the impression that the `ptr` value here must be freed via a call to `ct_status_free` - is that right? or is this just saying that the lifetime of the returned result must be shorter than `ContractStatus`?

If it's the former (the `ptr` value must be freed explicitly): Are we leaking?
If it's the latter (the `ptr` values lives as long as `ContractStatus`): Is this code unsound? Would it be possible to keep using the returned `CString` after `self` has been dropped?

Reply:
The ptr returned is tied to the lifetime of the contract status handle `ct_stathdl_t`.  We read that pointer as a cstr but the function returns a CString which copies the data to own it. If I understand correctly the ptr returned from the call to `ct_pr_status_get_svc_fmri` will not leave anything around because it will be cleaned up when `ct_status_free` is called via the drop method on `ContractStatus`.  

I will verify this, but this was at least my intention upon writing this code.

Reply:
Ah, gotcha, so:

```
        let cstr = unsafe { CStr::from_ptr(ptr) }
```

Would create a `cstr` that needs to have a shorter lifetime than the handle, but:

```
        cstr.to_owned()
```

Would clone it, and free us from that lifetime constraint when the function returns